### DISCUSSION TITLE
Changing the values of the fluid's critical point properties predefined in the CoolProp

### DISCUSSION THREAD

**Discussion by mojtaba136411** [CoolProp/CoolProp#2316] [General]
I want to change the values of the fluids critical point properties predefined in the coolprop. I have installed it in MATLAB.
For instance the critical Temp for R245 is 427.01, I want to change it to 427.2.
How can I do this?
bests

**Comment by ibell** [comment]
Intentionally you cannot easily make this change. First off, why do you
want to make this change?

In python the process is relatively simple: you load the JSON from CoolProp
at runtime, change the reducing value, and reload the fluid. I am sure you
can do something similar in MATLAB. Would the Python example help?

**Reply by mojtaba136411** [reply]
uncertainty propagation of critical Temperature at the models.
Tc=[426.8-427.1]

**Reply by mojtaba136411** [reply]
is it correct??for example:

Tc_coolprop = py.CoolProp.CoolProp.PropsSI('Tcrit','Q',0.0,'T',273,'R245fa');

Tc_sampling_uncertainty = 426.95;

Reducing_value = Tc_coolprop/Tc_sampling_uncertainty

h = py.CoolProp.CoolProp.PropsSI('H','T',373*Reducing_value,'Q',0,'R245fa');

**Comment by ibell** [comment]
No, it is more complicated than that. You need to obtain the JSON for the fluid, change the critical temperature everywhere in the JSON file, and reload it, and do that for each critical temperature you want to set.

**Comment by mojtaba136411** [comment]
Where is the JSON file on the PC?
I don't find that in this path:
C:\Users\Name\miniconda3\Lib\site-packages\CoolProp

**Reply by henningjp** [reply]
Try this example on the CoolProp site.

[Adding a Fluid](http://www.coolprop.org/coolprop/HighLevelAPI.html#adding-fluids)

Hope this helps.

**Comment by mojtaba136411** [comment]
thanks Jeff

**Comment by ibell** [comment]
The process works like this as an example of changing the critical temperature for argon +- 1 K:

```python
import json 
import CoolProp.CoolProp as CP 
import numpy as np

fluid = 'Argon'
AS = CP.AbstractState('HEOS', fluid)
CP.set_config_bool(CP.OVERWRITE_FLUIDS, True)

# Get the fluid information in JSON format 
# for loaded fluid
jbase = json.loads(AS.fluid_param_string("JSON"))[0]

for Tc in np.linspace(150.687-1, 150.687+1, 7):
    # Change the max temperature in ancillary functions
    for k in ['pS', 'rhoL', 'rhoV']:
        jbase["ANCILLARIES"][k]["T_r"] = Tc
        jbase["ANCILLARIES"][k]["Tmax"] = Tc
    
    # Change the reducing temperature for EOS and
    # critical temperature at top-level
    jbase["EOS"][0]["STATES"]['reducing']['T'] = Tc
    jbase["STATES"]['critical']['T'] = Tc
    
    CP.add_fluids_as_JSON('HEOS', json.dumps(jbase))
    
    print(Tc, CP.PropsSI('T','P',101325,'Q',0,fluid))
```

So the JSON data structure is loaded in memory, the necessary coefficients are perturbed, and the perturbed data structure is reset in CoolProp.

I should warn you this is an advanced use case, so you should definitely sanity check that the results are working as expected. On my side, the code yielded:
```
149.687 86.78545291004171
150.02033333333335 86.95771642184718
150.35366666666667 87.1299441634586
150.687 87.3021362301716
151.02033333333335 87.4742927168127
151.35366666666667 87.64641371774503
151.687 87.81849932687082
```
which seems plausible

**Comment by mojtaba136411** [comment]
Thanks
that is goood

**Comment by mojtaba136411** [comment]
Hi Ian
Which folder should I put this code in?
test?
thanks alot

**Comment by ibell** [comment]
It shouldn't matter,  everything is done in memory.

**Comment by mojtaba136411** [comment]
Ok, thanks.

**Comment by mojtaba136411** [comment]
Dear Ian
please
please
please
Run the upper coments at the MATLAM envi.

**Comment by ibell** [comment]
Sorry I don't have MATLAB

**Comment by mojtaba136411** [comment]
:((

**Comment by mojtaba136411** [comment]
Hi Ian
I rewrote your code as follows:

import json
import CoolProp
import CoolProp.CoolProp as CP

fluid = 'Argon'

AS = CP.AbstractState('HEOS', fluid)
jbase = json.loads(AS.fluid_param_string("JSON"))[0]
# print(jbase)

Tc=149
# Change the max temperature in ancillary functions
for k in ['pS', 'rhoL', 'rhoV']:
    jbase["ANCILLARIES"][k]["T_r"] = Tc
    jbase["ANCILLARIES"][k]["Tmax"] = Tc
    jbase["EOS"][0]["STATES"]['reducing']['T'] = Tc
    jbase["STATES"]['critical']['T'] = Tc

CP.set_config_bool(CP.OVERWRITE_FLUIDS, True)
CP.add_fluids_as_JSON('HEOS', json.dumps(jbase))

print(CP.PropsSI('T','P',101325,'Q',0,fluid))

why don't HEOS add to the dev-Fluids folder?
Why dont see the output from (CP.PropsSI('T','P',101325,'Q',0,fluid)?

**Comment by mojtaba136411** [comment]
Hi Dear Ian
Finally, I was able to change the fluids parameters in CoolProp with MATLAB.
Thanks for your help.
function out_CP = Coolprop_Change_Param(T_crit,M_W_R)

% main value : Tc = 427.01 MWR = 134.04794
fluid = 'R245fa';

% jj is string: py.str
jj = py.CoolProp.CoolProp.get_fluid_param_string(fluid, 'JSON');
py.CoolProp.CoolProp.set_config_bool(py.CoolProp.CoolProp.OVERWRITE_FLUIDS,
true);

% jbase_1 is list: py.list
jbase_1 = py.json.loads(jj);

% jbase is dictionary: py.dict
jbase = jbase_1{1};

Tc = T_crit;
ancillaries = {'pS', 'rhoL', 'rhoV'};
% Change the max temperature in ancillary functions
for i = 1:length(ancillaries)
    k = ancillaries{i};
    jbase{'ANCILLARIES'}{k}{'T_r'} = Tc;
    jbase{'ANCILLARIES'}{k}{'Tmax'} = Tc;
end
% Change the reducing temperature for EOS and critical temperature at
top-level
jbase{('EOS')}{1}{'STATES'}{'reducing'}{'T'} = Tc;
jbase{'STATES'}{'critical'}{'T'} = Tc;

**Comment by mojtaba136411** [comment]
Hi Dear Jeff
Finally, I was able to change the fluids parameters in CoolProp with MATLAB.
Thanks for your help.

function out_CP = Coolprop_Change_Param(T_crit,M_W_R)

% main value : Tc = 427.01 MWR = 134.04794
fluid = 'R245fa';

% jj is string: py.str
jj = py.CoolProp.CoolProp.get_fluid_param_string(fluid, 'JSON');
py.CoolProp.CoolProp.set_config_bool(py.CoolProp.CoolProp.OVERWRITE_FLUIDS,
true);

% jbase_1 is list: py.list
jbase_1 = py.json.loads(jj);

% jbase is dictionary: py.dict
jbase = jbase_1{1};

Tc = T_crit;
ancillaries = {'pS', 'rhoL', 'rhoV'};
% Change the max temperature in ancillary functions
for i = 1:length(ancillaries)
    k = ancillaries{i};
    jbase{'ANCILLARIES'}{k}{'T_r'} = Tc;
    jbase{'ANCILLARIES'}{k}{'Tmax'} = Tc;
end
% Change the reducing temperature for EOS and critical temperature at
top-level
jbase{('EOS')}{1}{'STATES'}{'reducing'}{'T'} = Tc;
jbase{'STATES'}{'critical'}{'T'} = Tc;
